src/main/java/org/apache/sling/api/resource/AbstractStoppableDepthFirstResourceVisitor.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.sling.api.resource;
+
+import java.util.Iterator;
+
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * This visitor will traverse the given resource and all its children in a depth-first approach
+ * and call the {@link AbstractStoppableDepthFirstResourceVisitor#visit(Resource)} method for each visited resource.
+ * It decouples the actual traversal code from application code in its visit method. 
+ * 
+ * Concrete subclasses must implement only the {@link AbstractStoppableDepthFirstResourceVisitor#visit(Resource)} method.
+ * 
+ * The difference between this class and {@link AbstractResourceVisitor} is that this class allows to stop
+ * the traversal early or skip some parts of the resource tree.
+ *
+ * @see <a href="https://en.wikipedia.org/wiki/Depth-first_search">Depth-First-Search</a>
+ * @since 2.12.0 (Sling API Bundle 2.22.0)
+ */
+public abstract class AbstractStoppableDepthFirstResourceVisitor {
+
+    /**
+     * Defines how the traversal should be continued.
+     * <ul>
+     * <li>{@link #NORMAL} for regular continuation</li>
+     * <li>{@link #SKIP_SUBTREE} to skip the children of the current resource and continue with its siblings</li>
+     * <li>{@link #STOP} to stop the whole traversal (i.e. to no visit any other resource afterwards)</li>
+     */
+    enum TraversalContinuation {
+        NORMAL,
```

Comment:
I would like to propose a set of alternate names. For your consideration only, I don't necessarily view them as "better"

- CONTINUE
- SKIP_CHILDREN
- SKIP_ALL

They sound a bit more natural to me, however it's naming so things are very subjective here.